### ISSUE TITLE
Speedups not available in shapely 1.8.0 with Python 3.10

### ISSUE THREAD

**Issue by jwhitlock** [shapely/shapely#1263]
In Python 3.9.9, `speedups.available` is `True`, but it is `False` with Python 3.10.1.

I'm using this Dockerfile to test:

```Dockerfile
FROM python:3.10.1-slim

RUN apt-get update && \
    apt-get -y install libgeos-dev
RUN pip install shapely==1.8.0
RUN python -c "import sys; from shapely import speedups; sys.exit(not speedups.available)"
```

``docker build .`` fails on 3.10.1 on my two systems:

* macOS Big Sur (11.6), Intel x86_64
* macOS Monterey (12.1), Apple M1 arm64

If I change to ``python:3.9.9-slim``, both systems pass.

**Comment by jorisvandenbossche** [comment]
The reason might be that there are not (yet) wheels available for Python 3.10 for Mac (https://pypi.org/project/Shapely/#files only lists 3.10 wheels for manylinux). 
(for windows there is https://github.com/shapely/shapely-wheels/pull/10, don't directly see an issue for Mac)

So your pip install is doing a source install, in which case it might be needed to have numpy installed in advance to build the speed-ups.

**Comment by jwhitlock** [comment]
This is a Docker build, and Docker for the Mac runs in a Linux VM.

Running with ``docker build . --no-cache --progress=plain``, partial output, shows a wheel being used:

```
#5 [3/4] RUN pip install shapely==1.8.0
#5 sha256:6b24d9f47cd4ab648c6d151343674b87a5aecdea62d702897328a15b70ef3532
#5 2.125 Collecting shapely==1.8.0
#5 2.216   Downloading Shapely-1.8.0-cp310-cp310-manylinux_2_17_aarch64.manylinux2014_aarch64.whl (1.1 MB)
#5 2.362 Installing collected packages: shapely
#5 2.495 Successfully installed shapely-1.8.0
#5 2.496 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#5 2.684 WARNING: You are using pip version 21.2.4; however, version 21.3.1 is available.
#5 2.684 You should consider upgrading via the '/usr/local/bin/python -m pip install --upgrade pip' command.
#5 DONE 3.0s
```

I'm curious if Linux users have a similar experience.

**Comment by jwhitlock** [comment]
Here's a version of Dockerfile that forces source:

```
FROM python:3.10.1-slim

RUN apt-get update && \
    apt-get -y install libgeos-dev
RUN pip install numpy cython
RUN pip install --no-binary=shapely shapely==1.8.0
RUN python -c "import sys; from shapely import speedups; sys.exit(not speedups.available)"
```

For me, it installed (from wheels): 
* cython 0.29.26
* numpy 1.21.4

but the speedup check still failed. Here's the full output:

https://gist.github.com/jwhitlock/8d4d3441c6eb07d145db7f4953238f68

**Comment by mwtoews** [comment]
I can replicate the error. You'll find more info with `pip install --verbose ...` like "command 'gcc' failed: No such file or directory", which is fixed with (e.g.) `apt-get install build-essential`.

But this will also fail because the source distribution contains cythonized C files from an old version that is broken with Python 3.10. Perhaps #1264 might help in preventing these C files from being packed with the next source distribution.

**Comment by jorisvandenbossche** [comment]
And I can also replicate that the linux wheels don't have the speedups properly available. In a clean environment with just python 3.10:

```
(test-shapely-wheel) $ pip install shapely
Collecting shapely
  Downloading Shapely-1.8.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.2 MB)
     |████████████████████████████████| 1.2 MB 1.9 MB/s            
Installing collected packages: shapely
Successfully installed shapely-1.8.0
(test-shapely-wheel) $ python
Python 3.10.1 | packaged by conda-forge | (main, Dec 16 2021, 17:57:26) [GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import shapely.speedups
>>> shapely.speedups.available
False
```

It gives this error:

```
>>> from shapely.speedups import _speedups
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: /home/joris/miniconda3/envs/test-shapely-wheel/lib/python3.10/site-packages/shapely/speedups/_speedups.cpython-310-x86_64-linux-gnu.so: undefined symbol: _PyGen_Send
```

which seem the same issue as what was fixed in https://github.com/shapely/shapely/pull/1224 (so that was not only for Windows)

**Comment by jwhitlock** [comment]
The current main branch works for me:

```
FROM python:3.10.1-slim

RUN apt-get update && \
    apt-get -y install git libgeos-dev build-essential
RUN pip install \
    numpy==1.21.4 \
    cython==0.29.25
RUN pip install --verbose --no-binary=shapely \
    "git+https://github.com/shapely/shapely@907c7ee19219b74c45e032f085d3dd3020e0544d#egg=shapely"
RUN python -c "import sys; from shapely import speedups; sys.exit(not speedups.available)"
```

Full output at https://gist.github.com/jwhitlock/bddbd0fe2434eae86262770cc6f9cc23

**Comment by jorisvandenbossche** [comment]
> The current main branch works for me:

The main branch has been completely refactored compared to 1.8, so that's not a good comparison (but good to see that it builds correctly, of course). You could check the `maint-1.8` branch though, which should have the cython fix from https://github.com/shapely/shapely/pull/1224

In the main branch, the speedups are essentially always available (in a different form), so the `speedups.available` attribute to simply set to True (kept for backwards compatibility):

https://github.com/shapely/shapely/blob/907c7ee19219b74c45e032f085d3dd3020e0544d/shapely/speedups.py#L7

**Comment by jorisvandenbossche** [comment]
Thanks for testing! Good to see the maint-1.8 branch is working properly.

Small additional note: I think you don't need to install cython manually up front, as the `pip install` does that for you in an isolated build environment (you probably still want to install numpy manually, though, as otherwise the speedups are available (because built), but not necessarily used to the full extent).

**Comment by sgillies** [comment]
This will be fixed in 1.8.1.
